### DISCUSSION TITLE
Using the mouse to resize splits?

### DISCUSSION THREAD

**Discussion by Erotemic** [kovidgoyal/kitty#5959] [Ideas]
Hi, 

I'm aware that I can use ctrl+r and S or T to make the window shorter or taller for a horizontal split, but one of the things I currently value greatly in my current terminal (terminator) is the ability to drag to resize splits.

I'm wondering if:

1. There currently exists any way to use the mouse click + drag to resize splits
2. If there is any interest in this feature or if there has been design decision to explicitly not support this feature.

**Comment by kovidgoyal** [comment]
1. no 2. patches welcome I just dont care enough for this to implement it myself

**Comment by pcewing** [comment]
I just started using Kitty this week and this is a feature I really want and am also interesting in adding. I use the `mod+right click drag` in i3wm to resize windows and it's really handy.

I made a little PoC of this earlier via a `mouse_resize` kitten:
![kitty_mouse_resize](https://user-images.githubusercontent.com/11671852/218195790-ebd7dc08-66f5-4ea3-8efc-964e60fc57ea.gif)

I pushed that to my branch in case you want to see what that looks like in code:
https://github.com/pcewing/kitty/tree/mouse-resize

It's still full of bugs but just meant to prove that this could be possible.

I think coming up with a full design for this that would be merge-able will require a bit of thought as there are a handful of cases to consider; however, i3wm has implemented this quite well so I think in many cases their design could be used.

Some thoughts/challenges:
- Behavior will sometimes need to change based on where the initial mouse button press occurs in the window
    - For a window that is horizontally located between adjacent windows on both left and right, clicking in the left half should cause the resize mechanism to move the left boundary and vice versa for the right
    - The same above rule needs to apply but vertically
- How to terminate the kitten may be tricky
    - `mouse_map` doesn't handle button draggin to my knowledge so in my PoC I just map the kitten starting to `ctrl+right click` but theoretically a different mouse button could be used
    - In the kitten, it needs to figure out which mouse button was pressed to launch the kitten so that it can stop executing when that same button is released
- Another reason terminating the kitten is tricky is that mouse events are only sent to windows underneath the cursor
    - If the termination condition is the mouse button being released, but the cursor was dragged outside of the window boundaries, the kitten won't get that mouse event

There are probably solves to those problems above but just some observations from the time I spent putting together the PoC. I'm going to keep hacking on this in my free time

**Comment by pcewing** [comment]
With a 3 month old baby haven't had much free time recently but I noodled on this a little more, trying instead to implement this "natively" in C or Python instead of a kitten. And rather than holding a modifier and clicking and dragging anywhere in a window, implementing it such that the user could just click and drag on the border line between windows.

It's somewhat functional but still quite buggy:
![Recording 2023-02-16 at 22 36 05](https://user-images.githubusercontent.com/11671852/219558587-071ecf75-21e1-4288-b29f-8dbd937b00b8.gif)

Border lines don't get redrawn and sometimes ctrl+L needs to be pressed to redraw a window and fix the text in it. Also I only implemented clicking between two windows as a PoC but didn't worry about more complex cases like clicking on a border that joins 3 or 4 windows.

Going to dig into what all the `kitty/layout/splits.py:Splits.modify_size_of_window()` function does as that's what the `resize_window` kitten ultimately uses from what I can tell.

One note is that this could be complex to try and implement for all of the various layouts. Given that I only ever use the Splits layout that's all I'm probably going to worry about at least for now.

**Comment by kovidgoyal** [comment]
The resize_window kitten uses remote control, the kitty @ resize_window
command, which in turn calls Boss.resize_layout_window()

**Reply by pcewing** [reply]
Thanks! Yea I was able to figure out that this is the basic callstack for resizing via the kitten:
```
kittens/resize_window/main.py:Resize.do_window_resize()
kitty/rc/resize_window.py:ResizeWindow.message_to_kitty()
kitty/rc/resize_window.py:ResizeWindow.response_from_kitty()
kitty/boss.py:Boss.resize_layout_window()
kitty/tabs.py:Tab.resize_window_by()
kitty/layout/splits.py:Splits.modify_size_of_window()
```

My little poc is basically just setting window geometry so I'll look through what else happens in the above to get the borders to redraw, etc.

**Comment by pcewing** [comment]
So I played around with this a bit more and got it into a pretty good spot:
![Recording 2023-02-19 at 16 10 20](https://user-images.githubusercontent.com/11671852/219981892-f0f19294-aca1-40c4-926e-19ebd4c1063f.gif)

Didn't want to record/upload a several minute gif so that only shows a subset of the possible cases but I believe I'm handling all normal cases now:
- Dragging a border that joins two windows
- Dragging the intersection of borders that join three windows
- Dragging the intersection of borders that join four windows

There are some assumptions being made that I'm not sure are correct, such as assuming that border width/height are always the same as cell width and height. From reading some of the possible config options, I don't think that's true.

There's also some room for improvement in general and handling edge cases , some of which is called out in TODOs. I think changing the cursor when hovering above borders would be a nice feature as it can be hard to tell whether or not you're clicking on a border. Also, highlighting the borders that are being adjusted when the mouse button is held. But, as a general draft I'd say I'm pretty happy with the direction so far.

Code is in my `mouse-resize-2` branch:
https://github.com/kovidgoyal/kitty/compare/master...pcewing:kitty:mouse-resize-2

Going to continue solidifying this and working towards getting it in a more mergeable state.

**Reply by Erotemic** [reply]
This looks really nice! It would be really cool to have this as a feature!

**Reply by Erotemic** [reply]
Letting you know I did attempt to build this and I was getting an error:

```
Linking [wayland] kitty/glfw-wayland ...
lto1: fatal error: bytecode stream in file ‘build/glfw-wayland-monotonic.c.o’ generated with LTO version 9.4 instead of the expected 11.3
```

I'm on Ubuntu 22.04 and running X11. Not sure if this is wayland only for this POC, but thought I'd provide the feedback.

**Reply by pcewing** [reply]
Sorry was traveling last week. Are you able to build the `master` branch without hitting that? I'm not sure why any of the changes I made would cause that and I'm also running X11 (i3wm) on Ubuntu 22.04. That branch is pretty old at this point so maybe it just needs to be rebased on latest `master`.

**Comment by kovidgoyal** [comment]
Looks good. Some minor comments from looking at the GIF only.

1) you will need to implement it for all layouts to be mergeable.

2) given that borders are so thin I am guessing (without looking at the code)
that there is some zone around the borders in which one can click to
drag. This is going to interfere with mouse selection/clicking on
hyperlink etc. It might be better to have a dedicated modifier for this.
So for example, holding shift+meta+left mouse down will start the resize. This
will need to be configurable via the existing mouse_actions framework.
I may have misunderstood the GIF and you are doing this already, if so
apologies.

**Comment by pcewing** [comment]
I appreciate the feedback!

I wanted to get this working just for the splits layout initially since that's the only layout I use but now that I know I can take advantage of `Boss.resize_layout_window()` it should be pretty simple to adapt to other layouts. The most complicated part of implementing the border drag resizing was just identifying which window(s) to resize depending on the position of the border being dragged.

My most recent gif shows the ability to click and drag on window borders to resize but my original post worked more like what you are saying. It mimicked the behavior of i3wm/swaywm where the user can hold a modifier and then click and drag anywhere in a window to resize that window.

Now that I've had some time to familiarize with the codebase (This is my first contribution) I think I have a better idea of how things work. I'll probably go back to that original idea because I prefer the resizing behavior in i3wm and like you said, trying to click narrow borders is a bit of a pain. That said, now that I know how, I'll implement it without adding a kitten which introduced a few complications.

Will follow up once I have something that's less of a proof-of-concept and closer to being mergeable.

**Comment by kovidgoyal** [comment]
Cool, all the best. I look forward to your PR.

**Comment by pcewing** [comment]
Haven't forgotten about this but been busy with baby lately. Had some time today so I gave this another pass. Changed it back to being more like the mouse resize functionality in `i3wm`. While holding `ctrl`, clicking and dragging in a window will resize. I tested all of the cases I could think of in the splits layout and think I have them covered:

![Recording 2023-03-16 at 19 21 10](https://user-images.githubusercontent.com/11671852/225788127-b03ad5c3-1350-405f-a8a7-1534baf3ed45.gif)

Theoretically I think it should work for all layouts but I could be wrong so I need to spend some time testing all of them. Also need to add a bit of polish:
- Add the ability to enable/disable the feature via configuration
- Make the modifier and/or mouse button configurable (Currently hard-coded to `ctrl + left click`)
- Change the cursor while resizing
    - Not sure how feasible this is but would be nice to change to something like the below
- Clear windows after resizing
    - As shown in the gif above, the output gets jumbled sometimes when resizing
    - This behavior is present in the current keyboard resizing but it's rather annoying
    - Would be nice to do the equivalent of a `ctrl + L` in each window on the active tab
    - AFAIK `ctrl + L` is a bash shortcut, not sure if there's a shell-agnostic way for kitty to do that

![resize_cursor](https://user-images.githubusercontent.com/11671852/225788466-3c9bf440-0054-4d6b-a4de-662a56422b93.png)

Hopefully will have a PR up sometime soon but if anyone wants to help with testing or just play around with this, code is in my `mouse-drag-resize` branch:
https://github.com/kovidgoyal/kitty/compare/master...pcewing:kitty:mouse-drag-resize

**Reply by Erotemic** [reply]
I see the appeal of holding ctrl, but the feature looses a lot of appeal to me if I can't do it with one hand by clicking the separator. One of my PCs is in my living room in a standing location and sometimes the keyboard is out of reach or unplugged and I just want to check status of something on my terminal. That might involve resizing it to see more output. I've been in this case more than once.

Perhaps this doesn't matter to most people, but I thought I'd share the feedback. Thank you for working on this feature in any capacity.

**Reply by pcewing** [reply]
Understandable, I think both implementations have pros and cons. Clicking a narrow border can be difficult and isn't great for accessibility but at least it can be done with one hand. Using a modifier makes it easier to quickly click and drag but requires both hands.

I prototyped both ways and they were both pretty straight forward to implement. Could theoretically just add both but not sure if maintainers would want that

**Reply by Erotemic** [reply]
It might not be too bad to have the ability to just click and hold an edge (if you can find it), but if you are pressing control you can click from basically anywhere. But as you said, it's up to maintainers. I'd argue that both is the correct way to do it.

**Reply by kovidgoyal** [reply]
I am not a fan of clicking the edge to resie. Not only is it hard to do, it also means that clicks near the edge now become hit or miss. Since they could either start a selection or start a resize. As such I wont be accepting both.

**Reply by pcewing** [reply]
I do think those concerns could be alleviated by:
- Highlighting the border when mouse is hovering over it
- Changing cursor icon to a resize indicator when hovering over a border

Then it would be a bit easier to click accurately. That said, I don't feel strongly about having both methods and it introduces more code to maintain.

One other thought, does right click and drag do anything currently? For folks who don't want to have to hold a modifier, perhaps they could just make right click drag without a modifier do the resize.

**Reply by Erotemic** [reply]
I would be happy with any solution that doens't require a modifier key. To me, the whole point of using the mouse at all is that my hands aren't at my keyboard.

The way terminator implements this is by changing the mouse icon when a click will start a resize. The icon normally is this text selection icon: 

![image](https://user-images.githubusercontent.com/3186211/225964864-751559d3-8b1b-4365-9990-8ba370e20a6b.png) 

and when you hover over a horizontal or vertical separator it changes to an up/down or left/right icon: e.g. 

![image](https://user-images.githubusercontent.com/3186211/225965026-d0eb018d-179a-4100-b8d3-c46c6a86ccd9.png)

I expect text to be jumbled when I resize. There isn't much to do about that.

**Comment by kovidgoyal** [comment]
Another approach is to have modes. So when ctrl+shift+r particular modifier is
pressed kitty goes into window resize mode and shows the keyboard
controls as it does now, but additionally in this mode dragging with the
mouse resizes.

**Comment by kovidgoyal** [comment]
On Fri, Mar 17, 2023 at 09:30:13AM -0700, Paul Ewing wrote:
> One other thought, does right click and drag do anything currently? For folks who don't want to have to hold a modifier, perhaps they could just make right click drag without a modifier do the resize.

It extends selections. https://sw.kovidgoyal.net/kitty/overview/#mouse-features
and
https://sw.kovidgoyal.net/kitty/conf/#mouse-actions

**Comment by mazunki** [comment]
I was just looking to see if this was possible, and it looks like this isn't merged yet.

I realize @pcewing hasn't had much time with family and all, and I don't want to pressure them if they don't have time, but I was wondering if there were any updates on this. If you don't have time, or have abandoned the idea, I'd gladly pick it up and look into continuing your work.

I think it would be very reasonable to have this enabled by default in resize mode, but I agree with @Erotemic in that half the point of resizing with mouse is to avoid using the keyboard. In my case, my `window_margin_width` are 12pt in each direction, so aiming at the border in my case would be super easy.

One could maybe have some `resize_grab_width` which allows both positive and negative values. With positive values it states how near the border you have to be (which can overlap with text selection) to start a mouse resize, and with negative values it disables part of your margin as the resize-grab area. With a large `window_margin_width`, and the `resize_grab_width` set to `0`, the entire margin (but none of the contents of the window) would be the grab area. I'm thinking the default should be 1pt, but 🤷🏽 .

**Comment by fat-tire** [comment]
For this to have amazing value for me, a modifier key press would not be required.  Tilix as an example, but more generally speaking-- most resizeable windows are a simple select-and-drag to resize a split.  Sometimes I'll slide one window lower to look at something quickly (say, the status of a build) and then flip it up, squashing it mostly out of view (except for the last line or two), or do a resize to keep a part of different terminals in view -- if only for aesthetic reasons... So to have to hunt down some memorized key-combo at the same time and use multiple hands would be a real bummer and source of frustration.

All this said, the mouse resizing is essential so please do support a simple keyless select-and-drag.  I can bear the massive risk of accidentally misclicking an adjascent character (not a tragedy- this can happen when resizing windows too) and just having to try again to select the border. No big deal.

Thanks!

**Comment by robertkirkman** [comment]
Hello I like this so I rebased the patch on the current version of kitty it's still working for me. 
Also I wanted to let anyone reading know that there is a distro where you can install this and it doesn't matter that it's not merged because it only affects you but it still automatically reapplies the patch every time you get an update to kitty. It is Gentoo and in Gentoo I can just create the folder `/etc/portage/patches/x11-terms/kitty` and then put this patch file in there and then use `sudo emerge kitty` and every time I use `sudo emerge -avuDN world`  it will automatically install the newest Gentoo package of kitty for me but also reapply this patch so that everyone can get updates but also only get the mouse resizing patch if they actually need it.

<details>
<summary>mouse-drag-resize.diff</summary>

```diff
--- a/kitty/boss.py
+++ b/kitty/boss.py
@@ -13,6 +13,7 @@
 from contextlib import contextmanager, suppress
 from functools import partial
 from gettext import gettext as _
+from math import floor
 from gettext import ngettext
 from time import sleep
 from typing import (
@@ -374,6 +374,7 @@ def __init__(
         global_shortcuts: dict[str, SingleKey],
         talk_fd: int = -1,
     ):
+        self.drag_resize_active = False
         self.atexit = Atexit()
         set_layout_options(opts)
         self.clipboard = Clipboard()
@@ -2129,6 +2131,52 @@ def switch_focus_to(self, window_id: int) -> None:
         if tab:
             tab.set_active_window(window_id)
 
+    def drag_resize_start(self, x: float, y: float, cell_width: int, cell_height: int) -> None:
+        tab = self.active_tab
+        if not tab:
+            return
+
+        (horizontal, vertical) = tab.current_layout.drag_resize_target_windows(x, y, tab.windows)
+
+        self.drag_resize_cell_width = cell_width
+        self.drag_resize_cell_height = cell_height
+        self.drag_resize_target_horizontal = horizontal
+        self.drag_resize_target_vertical = vertical
+        self.drag_resize_active = True
+        self.drag_resize_initial_x = x
+        self.drag_resize_initial_y = y
+        self.drag_resize_last_step_x = 0
+        self.drag_resize_last_step_y = 0
+
+    def drag_resize_update(self, x: float, y: float) -> None:
+        if not self.drag_resize_active:
+            return
+
+        if self.drag_resize_target_horizontal is not None:
+            step_x = floor((x - self.drag_resize_initial_x) / self.drag_resize_cell_width)
+            dx = step_x - self.drag_resize_last_step_x
+            if not dx == 0:
+                self.resize_layout_window(self.drag_resize_target_horizontal, float(dx), True, False)
+                self.drag_resize_last_step_x = step_x
+
+        if self.drag_resize_target_vertical is not None:
+            step_y = floor((y - self.drag_resize_initial_y) / self.drag_resize_cell_height)
+            dy = step_y - self.drag_resize_last_step_y
+            if not dy == 0:
+                self.resize_layout_window(self.drag_resize_target_vertical, float(dy), False, False)
+                self.drag_resize_last_step_y = step_y
+
+    def drag_resize_end(self) -> None:
+        self.drag_resize_cell_width = 0
+        self.drag_resize_cell_height = 0
+        self.drag_resize_target_horizontal = None
+        self.drag_resize_target_vertical = None
+        self.drag_resize_active = False
+        self.drag_resize_initial_x = 0.0
+        self.drag_resize_initial_y = 0.0
+        self.drag_resize_last_step_x = 0
+        self.drag_resize_last_step_y = 0
+
     def open_kitty_website(self) -> None:
         self.open_url(website_url())
 
--- a/kitty/layout/base.py
+++ b/kitty/layout/base.py
@@ -448,6 +448,39 @@ def layout_state(self) -> dict[str, Any]:
     def set_layout_state(self, layout_state: dict[str, Any], map_group_id: WindowMapper) -> bool:
         return True
 
+    def drag_resize_target_windows(self, x: float, y: float, all_windows: WindowList) -> tuple[WindowType, WindowType]:
+        # Identify the window where the click occurred and which horizontal and
+        # vertical half it was in
+        click_window, left_half_clicked, top_half_clicked = (None, False, False)
+        for w in all_windows.all_windows:
+            g = w.geometry
+            if x >= g.left and x <= g.right and y >= g.top and y <= g.bottom:
+                click_window = w
+                left_half_clicked = g.left <= x and x <= g.left + (float(g.right - g.left) / 2.0)
+                top_half_clicked = g.top <= y and y <= g.top + (float(g.bottom - g.top) / 2.0)
+                break
+
+        if click_window is None:
+            raise Exception("Failed to determine click window")
+
+        neighbors = self.neighbors_for_window(click_window, all_windows)
+
+        # Infer which window should be horizontally resized based on click
+        # position and layout state
+        horizontal_target = click_window
+        if ((left_half_clicked and len(neighbors["left"]) > 0) or
+            (not left_half_clicked and len(neighbors["left"]) > 0 and len(neighbors["right"]) == 0)):
+            horizontal_target = all_windows.id_map[neighbors["left"][0]]
+
+        # Infer which window should be vertically resized based on click
+        # position and layout state
+        vertical_target = click_window
+        if ((top_half_clicked and len(neighbors["top"]) > 0) or
+            (not top_half_clicked and len(neighbors["top"]) > 0 and len(neighbors["bottom"]) == 0)):
+            vertical_target = all_windows.id_map[neighbors["top"][0]]
+
+        return (horizontal_target, vertical_target)
+
     def serialize(self, all_windows: WindowList) -> dict[str, Any]:
         ans = self.layout_state()
         ans['opts'] = self.layout_opts.serialized()
--- a/kitty/mouse.c
+++ b/kitty/mouse.c
@@ -660,6 +660,35 @@ closest_window_for_event(unsigned int *window_idx) {
     return ans;
 }
 
+static void
+drag_resize_start(double mouse_x, double mouse_y, unsigned int cell_width, unsigned int cell_height) {
+    call_boss(drag_resize_start, "ddII", mouse_x, mouse_y, cell_width, cell_height);
+}
+
+static void
+drag_resize_update(double mouse_x, double mouse_y) {
+    call_boss(drag_resize_update, "dd", mouse_x, mouse_y);
+}
+
+static void
+drag_resize_end(void) {
+    call_boss(drag_resize_end, "");
+}
+
+static bool
+is_in_window(double mouse_x, double mouse_y) {
+    Tab *active_tab = global_state.callback_os_window->tabs + global_state.callback_os_window->active_tab;
+
+    for (unsigned int i = 0; i < active_tab->num_windows; ++i) {
+        WindowGeometry *g = &active_tab->windows[i].render_data.geometry;
+        if (g->left <= mouse_x && mouse_x <= g->right && g->top <= mouse_y && mouse_y <= g->bottom) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
 void
 focus_in_event(void) {
     // Ensure that no URL is highlighted and the mouse cursor is in default shape
@@ -770,6 +799,27 @@ mouse_event(const int button, int modifiers, int action) {
     bool in_tab_bar;
     unsigned int window_idx = 0;
     Window *w = NULL;
+
+    OSWindow *osw = global_state.callback_os_window;
+
+    // Handle mouse drag window resizing
+    if (!global_state.active_drag_resize && button == GLFW_MOUSE_BUTTON_LEFT && action == GLFW_PRESS && modifiers & GLFW_MOD_CONTROL) {
+        if (is_in_window(osw->mouse_x, osw->mouse_y)) {
+            drag_resize_start(osw->mouse_x, osw->mouse_y, osw->fonts_data->fcm.cell_width, osw->fonts_data->fcm.cell_height);
+            global_state.active_drag_resize = true;
+            return;
+        }
+    } else if (global_state.active_drag_resize) {
+        if (button == GLFW_MOUSE_BUTTON_LEFT && action == GLFW_RELEASE) {
+            drag_resize_end();
+            global_state.active_drag_resize = false;
+            return;
+        } else if (button < 0) {
+            drag_resize_update(osw->mouse_x, osw->mouse_y);
+            return;
+        }
+    }
+
     if (OPT(debug_keyboard)) {
         if (button < 0) { debug("%s x: %.1f y: %.1f ", "\x1b[36mMove\x1b[m", global_state.callback_os_window->mouse_x, global_state.callback_os_window->mouse_y); }
         else { debug("%s mouse_button: %d %s", action == GLFW_RELEASE ? "\x1b[32mRelease\x1b[m" : "\x1b[31mPress\x1b[m", button, format_mods(modifiers)); }
--- a/kitty/state.h
+++ b/kitty/state.h
@@ -332,7 +332,7 @@ typedef struct {
     bool has_pending_resizes, has_pending_closes;
     bool check_for_active_animated_images;
     struct { double x, y; } default_dpi;
-    id_type active_drag_in_window, tracked_drag_in_window, mouse_hover_in_window;
+    id_type active_drag_in_window, tracked_drag_in_window, mouse_hover_in_window, active_drag_resize;
     int active_drag_button, tracked_drag_button;
     CloseRequest quit_request;
     bool redirect_mouse_handling;
```

</details>

**Reply by SharzyL** [reply]
> Hello I like this so I rebased the patch on the current version of kitty it's still working for me. Also I wanted to let anyone reading know that there is a distro where you can install this and it doesn't matter that it's not merged because it only affects you but it still automatically reapplies the patch every time you get an update to kitty. It is Gentoo and in Gentoo I can just create the folder `/etc/portage/patches/x11-terms/kitty` and then put this patch file in there and then use `sudo emerge kitty` and every time I use `sudo emerge -avuDN world` it will automatically install the newest Gentoo package of kitty for me but also reapply this patch so that everyone can get updates but also only get the mouse resizing patch if they actually need it.

Thank you and a Nix user is also enjoying it!

```nix
  kitty = prev.kitty.overrideAttrs (oldAttrs: {
    patches = (oldAttrs.patches or []) ++ [
      ./patches/kitty-mouse.patch
    ];
  });
```

**Reply by KatieFournier1** [reply]
Doesn't compile for me, so here's a patch for the patch!

<details>
<summary>kitty-mouse-patch.patch</summary>

```diff
--- a/patches/kitty-mouse.patch
+++ b/patches/kitty-mouse.patch
@@ -76,7 +76,7 @@
      def layout_state(self) -> Dict[str, Any]:
          return {}
 +
-+    def drag_resize_target_windows(self, x: float, y: float, all_windows: WindowList) -> Tuple[WindowType, WindowType]:
++    def drag_resize_target_windows(self, x: float, y: float, all_windows: WindowList) -> tuple[WindowType, WindowType]:
 +        # Identify the window where the click occurred and which horizontal and
 +        # vertical half it was in
 +        click_window, left_half_clicked, top_half_clicked = (None, False, False)
@@ -157,7 +157,7 @@
 +    // Handle mouse drag window resizing
 +    if (!global_state.active_drag_resize && button == GLFW_MOUSE_BUTTON_LEFT && action == GLFW_PRESS && modifiers & GLFW_MOD_CONTROL) {
 +        if (is_in_window(osw->mouse_x, osw->mouse_y)) {
-+            drag_resize_start(osw->mouse_x, osw->mouse_y, osw->fonts_data->cell_width, osw->fonts_data->cell_height);
++            drag_resize_start(osw->mouse_x, osw->mouse_y, osw->fonts_data->fcm.cell_width, osw->fonts_data->fcm.cell_height);
 +            global_state.active_drag_resize = true;
 +            return;
 +        }

```

</details>

**Reply by robertkirkman** [reply]
Thanks! I have now updated to Kitty version 0.42.2 and I have edited my original patch (which was originally based on pcewing's code) to be compatible with it.

**Reply by robertkirkman** [reply]
I have now updated to Kitty version 0.43.1 and I have rebased the patch on it past this commit! https://github.com/kovidgoyal/kitty/commit/d52f2e798164c2f2800eb28d227bead4f1b913a4 It seems like it's still working.

**Reply by robertkirkman** [reply]
I have now updated to Kitty version 0.46.2 which has an **official implementation of this**. I removed the patch and I began using the official implementation.

The official implementation has a different key binding. Instead of Ctrl + Shift + Mouse1 _anywhere_ + drag, it is now Mouse1 _on the green divider line specifically_ + drag.

I like it and if it continues to work for me then I will probably not try to go back to the old implementation.

**Comment by ssleert** [comment]
thx guys

**Comment by hexqubit** [comment]
Any updates?!

**Comment by juancampa** [comment]
Bumping this as it's a highly requested feature and feels like a low hanging fruit given the implementation is 99% there 🙏 

<img width="960" height="375" alt="Screenshot 2025-11-04 at 02 44 51" src="https://github.com/user-attachments/assets/090a8b9b-4058-4c51-b697-226c465d69be" />

**Reply by zzhaolei** [reply]
You can create your own patch version based on @pcewing’s commit and submit a PR.

**Comment by mcrmck** [comment]
I've submitted a PR implementing this feature: #9447

It's based on @pcewing's patch with a couple of additions:
- Fixed a KeyError when the layout returns a sparse neighbors map
- Added a move cursor during the drag operation                                                                                                                   

  Ctrl+left-click-drag on a window split to resize it.

**Reply by robertkirkman** [reply]
Thank you very much,

I am using the official version 0.46.2 which has your version in it and it's working very good for me and I think I can use it permanently now!
